First Pull

0xMoJo7 opened the pull request:
Everything functioning as expected.
What needs fixin'?

Commit:
Edit naming conventions and return lines

Change names of constant and datasets for style.
Return lines of movies and customers for readibility.

Commit:
Further reformatting

Added main function as well as moved constants under main.

Commit:
Chnage file names, Add enough_money func

Changed file names based on Coit's rec.
Also created the function to check if bank account for adequate funds

Commit:
Add deduct money function

Deduct $5 from Users account
Learned new testing method from Coit.

Commit:
Add refund function

Add $5 back to users account

Commit:
Change format, no driver function

Move database to test file
Edit enough_money test to pass

Commit:
Add enough_money function

Changed function to make cash an integer.

Commit:
Fix enough_money function

Actually pass test
Needed to pass library in test functions.

Commit:
Add charge function and test

Reverted overcomplicated function
Rewrite a test that made sense!

Commit:
Rework dispense ticket function, pass test

Overcomplicated it yet again!
Simple append and better written test did the trick.

Commit:
Add remove seat func and test, Reformat Movies DB

Altered MOVIE_TICKETS Library to be easier to call.
Found while writing remove_seat tests, should help me in the future.

Commit:
Push work done on airplane

Wrote out refund ticket functions.
Now having problems with string indices when removing a seat.
Will review git logs for issues but wanted to get code caught back up.

Commit:
Fix append method

I needed to change the way the title was passed.
Not necessary for "title" to be in []

Commit:
Pass Purchase Ticket main function

Tie all subfunctions back together.

Commit:
Fix remove ticket function and test

Followed dispense_ticket format

Commit:
Test refund ticket subfunctions

Got bit for copypasta on functions having remove_seats instead of
add_seat

Commit:
Change libraries for consistency in testing

Some still had [], other not, etc.

Commit:
Add pretty print module, add database print func

Use pretty print.
Inquired about testing methods.
Is pprint method clean enough?

Commit:
Add more elegant print functions

Remove pprint
Write better print functions.

Commit:
Create README.md

Commit:
Revise Readme

CoitThomas commented on code.py:
```diff
@@ -1,14 +1,26 @@
-def do_p(mv, mvg):
-    _mv_db[mv] = _mv_db[mv] - 1
-    if _mvg_d[mvg]["cash"] > 5.0:
+MOVIE_PRICE = 5.0
```
Move this to movies.py

CoitThomas commented on code.py:
```diff
@@ -1,14 +1,26 @@
-    if _mvg_d[mvg]["cash"] > 5.0:
+MOVIE_PRICE = 5.0
 
+def enough_money(customer):
```
move this to movies.py as well

CoitThomas commented on code.py:
```diff
@@ -1,14 +1,26 @@
-def tesst():
-    do_p("Xmen 8: The Xmennening", "Ricci")
-    assert _mv_db["Xmen 8: The Xmennening"] == 10 #WTF? 
+if __name__ == "__main__":
```
Nix this and subsequently the file - all of the Customer info and movie info would usually live in a database not in your code. Just use those dictionaries for testing purposes.

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,62 @@
+def enough_money(customer):
+    """Check if customer has adequate funds"""
+    balance = customer["cash"]
+    if balance < 5.0:
```
Once you move `MOVIE_PRICE = 5.0` into here, you can start referencing `MOVIE_PRICE` instead of hard coding `5.0` - this way, in the future, if the price of a movie changes, you only have to make 1 change instead of several.

CoitThomas replied:
It also makes things less "magical" - you know exactly what the 5.0 represents

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,62 @@
+
+def charge(customer):
+    """Charge customer $5"""
+    customer["cash"] = customer["cash"] - 5.0
```
MOVIE_PRICE

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,62 @@
+    return customer, movie
+
+def charge(customer):
+    """Charge customer $5"""
```
"""Charge the customer the price of 1 movie ticket"""

CoitThomas replied:
Be nice to your future self - that movie ticket price is probably going to change - you don't want future Eric having to hunt down outdated comments

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,62 @@
+
+def remove_seat(movie):
+    """Remove seat from ticket database"""
+    movie["seats"] = movie["seats"] - 1.0
```
Shouldn't this be an integer? You can't have fractions of a movie seat.

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,62 @@
+
+def remove_ticket(movie, customer):
+    """Removies ticket from customers library"""
+    customer["movies"].remove(movie["title"])
```
This looks like a great opportunity for an assertion. Assert that the customer actually has that movie you want to remove in their library before removing it. Go back through your other methods and be sure to utilize assertions in them as well.

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,62 @@
+
+def print_customer_data(customer):
+    """Prints customer database information"""
+    print("Movie Tickets:         ", customer["movies"])
```
The output of this will give you the movies in brackets: `[MovieA, MovieB, MovieC]` - try to find a cleaner way to present this.

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,62 @@
+    movie["seats"] = movie["seats"] + 1.0
+    return movie
+
+def print_movie_data(movie):
```
Add another print function that you can just feed an entire dictionary of movies to. You'll be able to utilize this function within the new one.

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,62 @@
+    print("Movie Title:           ", movie["title"])
+    print("Seats Remaining:       ", movie["seats"])
+
+def print_customer_data(customer):
```
Do the same thing here. Create another print function that can be fed an entire dictionary of customers.

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,111 @@
+    "Bromance" : {"title" : "The Bromance", "seats" : 20},
+    "Gigli" : {"title" : "Gigli: The Play: The Book: The movie", "seats" : 102}
+    }
+def test_enough_money():
```
New line

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,111 @@
+    jims_actual_cash = movies.charge(jim)["cash"]
+    jims_expected_cash = 5.0
+    assert jims_actual_cash == jims_expected_cash
+def test_dispense_ticket():
```
New line

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,111 @@
+    jims_library = movies.dispense_ticket(bromance, jim)["movies"]
+    jims_expected_library = ["Xmen 8: The Xmennening", "The Bromance"]
+    assert jims_library == jims_expected_library
+def test_remove_seat():
```
New line

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,111 @@
+    expected_bromance = {"title" : "The Bromance", "seats" : 19}
+    assert actual_bob == expected_bob
+    assert actual_bromance == expected_bromance
+def test_add_funds():
```
New line

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,111 @@
+    ricci_actual_cash = movies.add_funds(ricci)["cash"]
+    ricci_expected_cash = 9.0
+    assert ricci_actual_cash == ricci_expected_cash
+def test_remove_ticket():
```
New line

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,111 @@
+    cary_library = movies.remove_ticket(bromance, cary)["movies"]
+    cary_expected_library = ["Gigli: The Play: The Book: The movie"]
+    assert cary_library == cary_expected_library
+def test_add_seat():
```
New line

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,111 @@
+    }
+def test_enough_money():
+    """Tests the enough money function"""
+    bob = {"movies":[], "cash" : 100.0}
```
Instead of building a new customer, just use the ones that you already made in `CUSTOMERS`.
For example: `bob = CUSTOMERS["Bob"]`

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,111 @@
+def test_dispense_ticket():
+    """Test adding ticket to customers account"""
+    bob = {"movies": [], "cash": 100.0}
+    xmen = {"title" : "Xmen 8: The Xmennening", "seats" : 10}
```
Same thing with movies: `xmen = MOVIES["Xmen"]`

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,111 @@
+    "Ricci": {"movies": [],
+              "cash" : 4.0}
+}
+MOVIE_TICKETS = {
```
I think I told you to call this MOVIE_TICKETS at one point, but after seeing it again, I think MOVIES is probably a better choice

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,111 @@
+              "cash" : 4.0}
+}
+MOVIE_TICKETS = {
+    "Xmen" : {"title" : "Xmen 8: The Xmennening", "seats" : 10},
```
I would make "seats" a bit more descriptive. Something like "seats_available" or "tickets_remaining".

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,111 @@
+    """Tests the enough money function"""
+    bob = {"movies":[], "cash" : 100.0}
+    ricci = {"movies": [], "cash" : 4.0}
+    assert movies.enough_money(ricci) is False
```
Assertions of booleans do not require `is bool`. You can just do:
`assert not movies.enough_money(ricci)`
`assert movies.enough_money(bob)`

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,111 @@
+    bob = {"movies": [], "cash": 100.0}
+    bromance = {"title" : "The Bromance", "seats" : 20}
+    actual_bob, actual_bromance = movies.purchase_ticket(bromance, bob)
+    expected_bob = {"movies": ["The Bromance"], "cash" : 95.0}
```
Hard coding these numbers will be a big pain in the ass later. Try something along the lines of:
`expected_bob = {"movies": ["The Bromance"], "cash" : bob["cash"] - MOVIE_PRICE}`

CoitThomas reviewed (changes requested):
Despite the multiple comments, this is a good start. I will probably have more after you make these changes, but I figure this is enough for now. Let me know if you have any questions.

Commit:
Delete code.py file

No longer needed as reference.

Commit:
Change database to function

Change libraries to be under functions.
This way libraries are not altered in testing.

Commit:
Make Coit's upgrades from Pull Req

Still figuring out how to print all db
Figuring out how to print list without []

Commit:
Add print all of library functions

Used pretty print to return all raw data.

Commit:
Print customer movie lists without brackets

I needed to use the join function

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,100 @@
+
+MOVIE_PRICE = 5.0
+
+def customer_db():
```
This and the movie_db() function should really only be used in your testing module.

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,100 @@
+    print("Movie Tickets:         ", ",  ".join(customer["movies"]))
+    print("Account Balance:       ", customer["cash"])
+
+def print_all_movie_data():
```
Right now this function will only print the dictionary you've defined up above. This function and its sister function, print_all_customer_data(), should take in a dictionary of their respective data type. That will give these methods greater usability. You should probably also have some asserts in these functions to make sure they are the right type of dictionaries.

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,100 @@
+
+def charge(customer):
+    """Charge customer the price of one ticket"""
+    customer["cash"] = customer["cash"] - MOVIE_PRICE
```
I know it might seem redundant, but having functions which are as self-sufficient as possible can be pretty important, so I would assert that the customer has enough money in this method also. That way, let's say later you write a new feature that calls the charge() method,  you don't accidentally send people into debt. I would also add another test that captures this scenario.

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,134 @@
+    assert not movies.enough_money(ricci)
+    assert movies.enough_money(bob)
+
+def test_charge():
```
Just reiterating, there should be an assertion in here to make sure you aren't sending people into debt.

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,134 @@
+    jims_expected_library = ["Xmen 8: The Xmennening", "The Bromance"]
+    assert jims_library == jims_expected_library
+
+def test_remove_seat():
```
there should also be an assertion in here that makes sure you aren't removing seats which don't exist

CoitThomas replied:
So what I was getting at with this one, is that you should have a test in here similar to what you did here:
``` 
+    ricci = customers["Ricci"]
 +    assert not movies.enough_money(ricci)
```
create a move that has zero seats and then try to remove a seat from it

Commit:
Add assertions, print_all funcs

Add several more assertions.
Create loop for print_all funcs.

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,82 @@
+"""Movie ticket process system and database"""
```
Whoops, I missed this as well - no db stuff is in here

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,82 @@
+    """Performs the steps required in purchasing a ticket and returns altered customer and movie"""
+    assert enough_money(customer)
+    customer = charge(customer)
+    customer = dispense_ticket(movie, customer)
```
you can chain these two function calls on line 15 and 16:
`customer = dispense_ticket(movie, charge(customer))`

0xMoJo7 replied:
I get why this makes sense, and makes it shorter. But for the sake of readability --  wouldn't two lines be preferred?

CoitThomas replied:
This improves the readability by making it more concise.
i.e. - "dispense the ticket for the movie and charge the customer"

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,82 @@
+    """Charge customer the price of one ticket"""
+    assert enough_money(customer)
+    customer["cash"] = customer["cash"] - MOVIE_PRICE
+    return customer
```
let's also condense down line 23 and 24 to just:
`return customer["cash"] - MOVIE_PRICE`

CoitThomas replied:
ignore this - won't work because it's only altering a single field

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,82 @@
+    """Add movie to customer's account"""
+    assert movie["seats_avaliable"] >= 1
+    customer["movies"].append(movie["title"])
+    return customer
```
same here:
`return customer["movies"].append(movie["title"])`

CoitThomas replied:
ignore

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,82 @@
+    """Remove seat from ticket database"""
+    assert movie["seats_avaliable"] >= 1
+    movie["seats_avaliable"] = movie["seats_avaliable"] - 1
+    return movie
```
condense

CoitThomas replied:
ignore

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,82 @@
+def add_funds(customer):
+    """Add the value of one ticket to moviegoer's account"""
+    customer["cash"] = customer["cash"] + MOVIE_PRICE
+    return customer
```
condense

CoitThomas replied:
ignore

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,82 @@
+    """Removies ticket from customers library"""
+    assert movie["title"] in customer["movies"]
+    customer["movies"].remove(movie["title"])
+    return customer
```
condense

CoitThomas replied:
ignore

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,82 @@
+def add_seat(movie):
+    """Add seat back to database"""
+    movie["seats_avaliable"] = movie["seats_avaliable"] + 1
+    return movie
```
condense

CoitThomas replied:
ignore

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,140 @@
+    bobs_actual_cash = movies.charge(bob)["cash"]
+    bobs_expected_cash = 95.0
+    assert bobs_actual_cash == bobs_expected_cash
+    jim = customers["Jim"]
```
let's get some whitespace in here to separate out the tests within each test - you should go ahead and do this for all of your tests - otherwise, it can be a bit muddled and harder to follow what is actually being tested

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,82 @@
+def refund_ticket(movie, customer):
+    """Performs the steps required in refunding a ticket and returns altered customer and movie"""
+    customer = add_funds(customer)
+    customer = remove_ticket(movie, customer)
```
this one actually can be condensed

Commit:
Condense functions, add spacing

Condenses a few functions.
Added whitespace in testing for readability.

CoitThomas commented on movies_test.py:
```diff
@@ -0,0 +1,153 @@
+    jims_actual_cash = movies.charge(jim)["cash"]
+    jims_expected_cash = 5.0
+    assert jims_actual_cash == jims_expected_cash
+    ricci = customers["Ricci"]
```
give this some space too - it's testing something different

Commit:
Condensed more testing

Opportunities to make testing more short and sweet.

Commit:
More thorough testing of remove_seat func

Add edge cases that test with 1 or 0 seat(s)

Commit:
Add "breadcrumb" assertions

Make the tests more readable for outside viewer.
Do not need to scroll up to reference numbers, etc.

CoitThomas commented on movies.py:
```diff
@@ -0,0 +1,80 @@
+def enough_money(customer):
+    """Check if customer has adequate funds"""
+    balance = customer["cash"]
+    if balance < MOVIE_PRICE:
```
sorry, I should have seen this sooner, but if your logic is this simple, you can do some pretty sweet syntactic sugar in python: `return False if customer["cash"] < MOVIE_PRICE else True`

Commit:
Condense enough_money into one liner

Made update to cut 4 lines into 1 on enough_money func.

CoitThomas reviewed (approved):
🍄

Commit:
Modify print_all_movie_data for 100% pylint

Change unused variable to _

austinwiltshire commented on README.md:
```diff
@@ -0,0 +1,17 @@
+purchase_tickets contains the subfunctions charge, dispense_ticket, and remove_seat
+refund_tickets contains add_funds, remove_ticket, and add_seat.
+
+It is only necessary to use purchase_tickets and refund_tickets. The subfunctions should not be necessary on their own.
```
Break things into two files - purchase_tickets and refund_tickets. 

You can create a third file that has your shared globals - so purchase_tickets and refund_tickets will both import something like movie_database.py that has the globals like your price in them. 

All functions that you don't want to be used on their own, prepend them with a _, so add_funds, since it shouldn't be used on its own, should be renamed _add_funds

Python doesn't do this well, but we're getting closer and closer to 'object orientation' and what you need for add_funds, etc... that aren't necessary on their own is called "encapsulation". Other languages have sections called 'public' and 'private' for this.

We'll simulate that here.

austinwiltshire commented on movies.py:
```diff
@@ -0,0 +1,77 @@
+
+def enough_money(customer):
+    """Check if customer has adequate funds"""
+    return False if customer["cash"] < MOVIE_PRICE else True
```
you can shorten this to simply

```
return customer["cash"] >= MOVIE_PRICE
```

austinwiltshire commented on movies.py:
```diff
@@ -0,0 +1,77 @@
+    assert enough_money(customer)
+    customer = dispense_ticket(movie, charge(customer))
+    movie = remove_seat(movie)
+    return customer, movie
```
i like this returning of the variables.

you could probably put them in the same order as you took them in though for consistency

austinwiltshire commented on movies.py:
```diff
@@ -0,0 +1,77 @@
+
+def print_customer_data(customer): #movies have brackets around them
+    """Prints customer database information"""
+    print("Movie Tickets:         ", ",  ".join(customer["movies"]))
```
nice use of join

austinwiltshire commented on movies.py:
```diff
@@ -0,0 +1,77 @@
+
+def print_all_movie_data(movie_dictionary):
+    """Print all movie data raw"""
+    for _, details in movie_dictionary.items():
```
for dictionaries, you can iterate over just the values by using movie_dictionary.itervalues()

austinwiltshire commented on movies_test.py:
```diff
@@ -0,0 +1,160 @@
+
+import movies
+
+def customer_db():
```
i like these test helpers.

austinwiltshire commented on movies_test.py:
```diff
@@ -0,0 +1,160 @@
+    assert bob["cash"] == 100.0
+    assert movies.enough_money(bob)
+    bobs_actual_cash = movies.charge(bob)["cash"]
+    bobs_expected_cash = 95.0
```
really good use of names to make the code readable

austinwiltshire reviewed (changes requested):
few more structural and minor changes

Commit:
Removing movies.py and movie_test.py

Breaking out into purchase, refund, and print

Commit:
Break functions into purchase, refund, and audits

Additionally added details for purchase and refund
Added testing modules for each function.

Commit:
Fix docstrings for functions

Changed docstrings to be more explicit in adding Args and Returns.

Commit:
Small change to audit_databases

Catching up a few files as well.

Commit:
Update README.md file

Reflect more recent changes.

Commit:
Small change to audit_databases

Changed loop to show names with data.

austinwiltshire commented on test_refund_tickets_details.py:
```diff
@@ -0,0 +1,71 @@
+                  "cash" : 4.0}
+    }
+
+def movie_db():
```
refactor out this function and the customer db function into a test_helpers.py module and import that into each of these test files

Commit:
Remove duplicate database code

Import customer database instead of replicating.
